Add support for plugins that intercept open channel messages

remyers opened the pull request:
This PR introduces a new actor called `OpenChannelInterceptor` to intercept `OpenChannel` requests sent to a `Peer`actor from  the API (initiated) and remote peers (non-initiated).  

If a request is accepted, a `SpawnChannelInitiator` or  `SpawnChannelNonInitiator` message is sent back to the peer to continue the open channel flow. Otherwise an error message is sent back to the API call or remote peer. If a new request comes in while one is already being processed, it will return an error.

A new singleton actor called `PendingChannelsRateLimiter` is introduced to track pending channels and rate limit the creation of unconfirmed open channel request from remote peers. 

If the request is not rate limited and if an `InterceptOpenChannelPlugin` is defined, the plugin will be queried. The plugin must respond before a fixed timeout with either `AcceptOpenChannel` or `RejectOpenChannel`.  A subset of local channel parameters may also be modified when accepting channels. 

(Edit: updated scope of PR)

remyers commented:
@t-bast before I got further on this I wanted to check that you agree with my approach.  I like that it requires minimal changes to how a `Peer` handles requests to open a channel, but does require plugin developers respond with the previously internal `SpawnChannelNonInitiator` message.

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/PluginParams.scala:
```diff
@@ -55,4 +54,18 @@ trait CustomCommitmentsPlugin extends PluginParams {
+/**
+ * Each intercepted message type must be handled by at most one plugin.
+ */
+object InterceptedMessageType extends Enumeration {
+  type InterceptedMessageType = Value
+  val InterceptOpenChannel: InterceptedMessageType.Value = Value(1, "InterceptOpenChannel")
+  val InterceptUpdateAddHtlc: InterceptedMessageType.Value = Value(2, "InterceptUpdateAddHtlc")
+}
+
+trait InterceptMessagePlugin extends PluginParams {
+  // set of message types that can be intercepted by this plugin
+  def canIntercept: Set[InterceptedMessageType.Value]
```
This is premature factoring, it looks like you're trying to build a _generic_ way of intercepting arbitrary messages. That's not our goal since every interception mechanism will require very custom work and different mechanisms inside eclair to make them possible. The only known cases we'll want to provide interception for are channel opening and htlc forwarding.

Keep it simple, just define separate traits for each interception case:

```
trait InterceptChannelOpenPlugin extends PluginParams
(definitely not in this PR)
trait InterceptHtlcsPlugin extends PluginParams
```

t-bast reviewed (commented):
While this solution is simple, I think it would be better to have a more strictly defined data contract for channel open intercepting plugins. I think we need something like the following:

- the plugin exposes an `ActorRef[OpenChannelReceived]` and we directly send it to them (instead of using the event stream, since there is only one such plugin)
- the plugin's possible responses are well-defined in `PluginParams` and independent of the internal `Peer` classes: this will let us refactor `Peer` without breaking channel intercepting plugins
- in the `Peer`, when we need to defer the decision to a plugin, create a dedicated child actor for this that:
  - sends `OpenChannelReceived` to the plugin
  - waits for its response and translates it to the `Peer` data format (either `SpawnChannelNonInitiator` or an outgoing `Error`)
  - reject the channel open after a timeout (configurable in `eclair.conf`): we don't fully trust plugins to be correctly implemented, and we need to respond to our peer even if the plugin fails to tell us what to do

The plugin's response types defined in `PluginParams.scala` could look like:

```scala
sealed trait InterceptChannelOpenResponse
case class AcceptChannel(temporaryChannelId: ByteVector32, localParams: LocalParams, fundingAmount_opt: Option[Satoshi]) extends InterceptChannelOpenResponse
case class RejectChannel(temporaryChannelId: ByteVector32, error: Error) extends InterceptChannelOpenResponse
```

remyers commented:
I've incorporated the feedback from @t-bast to create a child actor with a defined UI to process messages between a channel interceptor plugin and the peer in [be6d270](https://github.com/ACINQ/eclair/pull/2552/commits/be6d270f412eecd155d30edac877962c73adebae). It still needs to be polished, but hopefully provides a better abstraction.

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/NodeParams.scala:
```diff
@@ -343,6 +345,9 @@ object NodeParams extends Logging {
     require(pluginFeatureSet.size == pluginMessageParams.size, "Duplicate plugin feature bits found")
 
+    val interceptOpenChannelPlugins = pluginParams.collect { case p: InterceptOpenChannelPlugin => p }
+    require(interceptOpenChannelPlugins.size <= 1, s"More than one plugin found that intercepts open channel messages: ${interceptOpenChannelPlugins.map(_.getClass.getSimpleName).mkString(", ")}")
```
nit: I'd give slightly more concrete steps for the node operator to fix the issue:
```suggestion
    require(interceptOpenChannelPlugins.size <= 1, s"At most one plugin is allowed to intercept channel open messages, but multiple such plugins were registered: ${interceptOpenChannelPlugins.map(_.getClass.getSimpleName).mkString(", ")}. Disable conflicting plugins and restart eclair.")
```

remyers replied:
fixed in [916da34](https://github.com/ACINQ/eclair/pull/2552/commits/916da349f31f465d4bb9b9dafec868da6cf8a34d)

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/PluginParams.scala:
```diff
@@ -55,4 +58,15 @@ trait CustomCommitmentsPlugin extends PluginParams {
+}
+
+// @formatter:off
+case class InterceptOpenChannelReceived(replyTo: ActorRef[InterceptOpenChannelResponse], open: Either[OpenChannel, OpenDualFundedChannel], temporaryChannelId: ByteVector32, localParams: LocalParams, fundingAmount: Satoshi)
```
nit: `fundingAmount` is redundant since it's already in `open_channel` and `open_channel2`? That could instead be a property of the case class:

```suggestion
case class InterceptOpenChannelReceived(replyTo: ActorRef[InterceptOpenChannelResponse], open: Either[OpenChannel, OpenDualFundedChannel], temporaryChannelId: ByteVector32, localParams: LocalParams) {
  val remoteFundingAmount: Satoshi = open match {
    case open: OpenChannel => open.fundingSatoshis
    case open: OpenDualFundedChannel => open.fundingAmount
  }
}
```

remyers replied:
Great suggestion! Fixed in [37eadd5](https://github.com/ACINQ/eclair/pull/2552/commits/37eadd56a1032d0c890abe0608d457df17496e4d).

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/PluginParams.scala:
```diff
@@ -55,4 +58,15 @@ trait CustomCommitmentsPlugin extends PluginParams {
+// @formatter:off
+case class InterceptOpenChannelReceived(replyTo: ActorRef[InterceptOpenChannelResponse], open: Either[OpenChannel, OpenDualFundedChannel], temporaryChannelId: ByteVector32, localParams: LocalParams, fundingAmount: Satoshi)
+sealed trait InterceptOpenChannelResponse
+case class AcceptOpenChannel(temporaryChannelId: ByteVector32, localParams: LocalParams) extends InterceptOpenChannelResponse
```
One of the important decisions that the plugin makes will be how much to contribute (when using dual funding), so we should probably add the necessary field right now to avoid breaking plugins later.

Also, more generally, why do we need to include `LocalParams` in `InterceptOpenChannelReceived` and `AcceptOpenChannel`? When looking at the fields inside this class, I don't think any of them are useful for the plugin, and I don't think the plugin should modify them, they directly come from the values the node operators chose in `eclair.conf` (and changing some of them would actually break the channel logic...)? I think I was the one to initially suggest passing `LocalParams`, but now that I look more into it I'm not sure it makes sense, unless you have a use-case for it?

```suggestion
case class AcceptOpenChannel(temporaryChannelId: ByteVector32, localFundingAmount: Satoshi) extends InterceptOpenChannelResponse
```

remyers replied:
I took a look at the channel interceptor support in CLN and LND to create a table of possible response fields for the Eclair channel interceptor: [channel_interceptor.md](https://gist.github.com/remyers/df02cb7eb938b29dc8715b7fa3098bac)

Perhaps you we can brainstorm around this topic a bit to make sure we have a comprehensive solution.

t-bast replied:
Good idea, I've answered in my top-level comment, there are some fields that we can indeed make configurable!

In the longer term, it may be interesting to also let those plugins do the channel funding part to let plugins implement things like vortex / coinjoin, but that will require more changes to the `Channel` logic so we'll wait to finalize dual funding / liquidity ads before adding that capability.

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/PeerEvents.scala:
```diff
@@ -18,6 +18,7 @@ package fr.acinq.eclair.io
 
 import akka.actor.ActorRef
 import fr.acinq.bitcoin.scalacompat.Crypto.PublicKey
+import fr.acinq.eclair.channel.{LocalParams, SupportedChannelType}
```
nit: remove changes to this file?

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala:
```diff
@@ -192,7 +193,7 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, wallet: OnChainA
         }
 
-      case Event(open: protocol.OpenDualFundedChannel, d: ConnectedData) =>
+        case Event(open: protocol.OpenDualFundedChannel, d: ConnectedData) =>
```
nit: the previous formatting was correct, this one isn't?

remyers replied:
Must have slipped in somehow, fixed in [ae99140](https://github.com/ACINQ/eclair/pull/2552/commits/ae991400f9cf8c3ca40869abf69cc8141ab8363e)

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala:
```diff
@@ -0,0 +1,75 @@
+
+import scala.concurrent.duration.FiniteDuration
+
+// Child actor of a Peer that handles the interactions with an InterceptOpenChannelPlugin:
+//  - sends OpenChannelReceived to the plugin
+//  - waits for its response and translates it to the Peer data format (either SpawnChannelNonInitiator or an outgoing Error)
+//  - reject the channel open after a timeout (configurable in eclair.conf)
+//
+//  Note: we don't fully trust plugins to be correctly implemented, and we need to respond to our peer even if the plugin fails to tell us what to do
```
nit: we usually put this kind of comment in the ScalaDoc of the actor, to make it automatically displayed in IDEs without having to go to the definition file. To do that, just format the comment with:

```scala
/**
 * Your comment
 */
```

And remove the newline between the comment and the `object`.

remyers replied:
Fixed in [8fc2a82](https://github.com/ACINQ/eclair/pull/2552/commits/8fc2a82f8712f007d263e773adc5da313fb75cc3). I'll add this to my checklist to avoid falling back into bad C++ commenting habits. :-)

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala:
```diff
@@ -0,0 +1,75 @@
+// Child actor of a Peer that handles the interactions with an InterceptOpenChannelPlugin:
+//  - sends OpenChannelReceived to the plugin
+//  - waits for its response and translates it to the Peer data format (either SpawnChannelNonInitiator or an outgoing Error)
+//  - reject the channel open after a timeout (configurable in eclair.conf)
```
nit: this isn't configurable anymore for simplicity :)

remyers replied:
Fixed in [7ff2a36](https://github.com/ACINQ/eclair/pull/2552/commits/7ff2a367a33eeddbe9d59b4b07dfc3a7f014b462)

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala:
```diff
@@ -0,0 +1,75 @@
+    }
+  }
+
+  // @formatter:off
+  sealed trait Command
+  case object PluginTimeout extends Command
+  case class WrappedOpenChannelResponse(pluginResponse: InterceptOpenChannelResponse) extends Command
+  // @formatter:on
```
very nit: we usually but `Command`s first in the object definition, since they're the "API" of the actor, most of the time they're the first thing we want to look at when reading the code.

remyers replied:
Fixed in [495fb2f](https://github.com/ACINQ/eclair/pull/2552/commits/495fb2feb79324a03ac4e19487ffa3c453bda863)

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/PluginParams.scala:
```diff
@@ -55,4 +58,15 @@ trait CustomCommitmentsPlugin extends PluginParams {
+// @formatter:on
+
+trait InterceptOpenChannelPlugin extends PluginParams {
+  def getOpenChannelInterceptor: ActorRef[InterceptOpenChannelReceived]
```
nit: getter/setter are so OOP, we usually omit them:
```suggestion
  def openChannelInterceptor: ActorRef[InterceptOpenChannelReceived]
```

remyers replied:
Fixed in [3c3a261](https://github.com/ACINQ/eclair/pull/2552/commits/3c3a2619e27a4ded6428ba3f587f1b9be223bff4)

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala:
```diff
@@ -0,0 +1,75 @@
+//  Note: we don't fully trust plugins to be correctly implemented, and we need to respond to our peer even if the plugin fails to tell us what to do
+
+object OpenChannelInterceptor {
+  def apply(replyTo: ActorRef[Any], plugin: InterceptOpenChannelPlugin, timeout: FiniteDuration, connectedData: ConnectedData, temporaryChannelId: ByteVector32, localParams: LocalParams, fundingAmount: Satoshi, open: Either[OpenChannel, OpenDualFundedChannel], channelType: SupportedChannelType): Behavior[Command] = {
```
We don't need the whole `ConnectedData`, just the `PeerConnection`. Also we don't need the `fundingAmount` if you apply changes from other comments:
```suggestion
  def apply(peer: ActorRef[Any], peerConnection: ActorRef[OutgoingMessage], plugin: InterceptOpenChannelPlugin, timeout: FiniteDuration, temporaryChannelId: ByteVector32, localParams: LocalParams, open: Either[OpenChannel, OpenDualFundedChannel], channelType: SupportedChannelType): Behavior[Command] = {
```

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala:
```diff
@@ -0,0 +1,75 @@
+        replyTo ! OutgoingMessage(Error(temporaryChannelId, "plugin timeout"), connectedData.peerConnection)
+        Behaviors.stopped
+      case WrappedOpenChannelResponse(a: AcceptOpenChannel) =>
+        replyTo ! SpawnChannelNonInitiator(open, ChannelConfig.standard, channelType, a.localParams)
```
The channel config should probably be provided to this actor, like we do for the channel type, otherwise if we introduce a different one than `standard` later we'll probably forget to update this piece of code.

remyers replied:
Fixed in [9c72ae8](https://github.com/ACINQ/eclair/pull/2552/commits/9c72ae850dee42e0e3ba5db1f205a8c8ceda46a9)

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala:
```diff
@@ -0,0 +1,75 @@
+    Behaviors.receiveMessage {
+      case PluginTimeout =>
+        context.log.error(s"plugin ${plugin.name} timed out while intercepting open channel")
+        replyTo ! OutgoingMessage(Error(temporaryChannelId, "plugin timeout"), connectedData.peerConnection)
```
There is a subtlety in the way you use the peer connection: the `Peer` will drop the message if the current connection doesn't match the one from this actor. This is actually ok, because if we disconnected before sending `accept_channel`, the channel opening flow is cancelled according to the lightning spec, so we shouldn't send those messages on the new connection. This is worth a comment!

remyers replied:
Should we also include `peerConnection` in the `SpawnChannelNonInitiator` message and update `Peer` to ignore it if `peerConnection` has changed?  

Otherwise I think we could be in a situation where the remote peer disconnects before we send `accept_channel` and according to the spec the remote peer must send a new `ChannelOpen` which we could accidentally reply to with an `accept_channel` from their previous `ChannelOpen` from before they disconnected.

remyers replied:
Updated comment in [3ee062b](https://github.com/ACINQ/eclair/pull/2552/commits/3ee062ba7cd78aa5e550e7917dd3709ea50f51a3). But may need to also update  how `SpawnChannelNonInitiator` is handled by `Peer` to live up to the comment.

t-bast reviewed (commented):
This is looking good on the architecture side, I have a lot of mostly cosmetic changes, but nothing too dramatic!

remyers commented:
@t-bast I think I've addressed the low hanging fixes, but two questions remain:
 - what [parameters](https://gist.github.com/remyers/df02cb7eb938b29dc8715b7fa3098bac) should be returned when we continue the open channel flow with `SpawnChannelNonInitiator`
 - should `SpawnChannelNonInitiator` include a `peerConnection` parameter and be ignored when the `peerConnection` changes after a peer reconnects

t-bast commented:
> what [parameters](https://gist.github.com/remyers/df02cb7eb938b29dc8715b7fa3098bac) should be returned when we continue the open channel flow with SpawnChannelNonInitiator

Interesting, there is indeed a subset of the parameters from `LocalParams` that we could let plugins set. I believe these parameters could be overridden by the plugin:

- dustLimit
- maxHtlcValueInFlightMsat
- htlcMinimum
- toSelfDelay
- maxAcceptedHtlcs

We could create a `case class DefaultLocalParams` that contains only this subset, pass it on to the plugin, and let them return a potentially updated one that we would then use to create our `LocalParams`.

In the future, we could also let the plugin do the channel funding part, which will let people build coinjoin/vortex-style transactions. But that will require deeper changes to the `Channel` actor (to skip the default funding flow), so we should defer that for now.

> should SpawnChannelNonInitiator include a peerConnection parameter and be ignored when the peerConnection changes after a peer reconnects

That's a very good point, if the connection changes while we're evaluating whether we want to accept a channel or not, we should ignore the result entirely! Well done noticing that.

I've been thinking more broadly about this PR and which checks should be done in plugins, and which should be in eclair-core. I think that there are some basic rate-limits that all plugins may want to apply, so it would make sense to implement them directly inside eclair-core to reduce duplication. What do you think about rate-limiting the number of pending channel requests directly inside the `OpenChannelInterceptor`?

I also feel like we should move **all** the logic of the channel validation inside the `OpenChannelInterceptor`: they clutter the `Peer` code, it would be cleaner if everything related to channel acceptance/configuration was done in `OpenChannelInterceptor`. It means this PR will move more code around, but the result will be a cleaner `Peer` actor, so I think it would be a win.

remyers commented:
Rebased on master; I'll see if I can add more tests and fixups, but otherwise ready to review.

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala:
```diff
@@ -0,0 +1,293 @@
+ * This simplifies the state machine and ensures that we do not introduce subtle race conditions when checking limits,
+ * since some of the conditions will be asynchronous (e.g. asking the router whether the node is public or private).
+ *
+ * When initiator (peer.OpenChannel):
+ *  - verify synchronous conditions against eclair.conf
+ *  - create LocalParams from nodeParams
+ *  - send SpawnChannelInitiator to parent Peer
+ *
+ * When non-initiator (protocol.OpenChannel, protocol.OpenDualFundedChannel):
+ *  - verify synchronous conditions against eclair.conf
+ *  - if the peer is not whitelisted to ignore rate-limits:
+ *    - ask the PendingChannelsRateLimiter if we should allow the channel
+ *    - if the rate limit is reached, reply to parent Peer with `OutgoingMessage(Error)`
+ *  - create LocalParams from nodeParams
+ *  - if a channel interceptor plugin is registered, ask it whether to allow the channel request:
+ *    - if the plugin rejects the request, reply to parent Peer with `OutgoingMessage(Error)`
+ *    - if the plugin doesn't respond in time, reply to parent Peer with `OutgoingMessage(Error)`
+ *    - if the plugin allows the request, update LocalParams with the plugin's response
+ *  - accept the channel, reply to parent Peer with `SpawnChannelNonInitiator`
```
I don't think it's a good idea to have such a detailed comment of the flow so far from the actual code. We will eventually update the code and will forget to update that comment, which will just become misleading. I'd rather remove that part and add comments on each behavior function to explain what it does when it's not obvious from the code itself.

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala:
```diff
@@ -0,0 +1,293 @@
+    }
+  }
+
+  private def checkRateLimits(nonInitiator: OpenChannelNonInitiator): Behavior[Command] = {
```
This should be the last thing we do, otherwise if the `PendingChannelsRateLimiter` accepts the channel but then we reject it for another reason, that channel will still use a slot in the `PendingChannelsRateLimiter` state that won't be freed, and at some point we'll end up rejecting all requests only because of such "zombie channels".

remyers replied:
Now that `PendingChannelRateLimiter` responds to `ChannelAborted` I think it makes sense to have `OpenChannelInterceptor` emit `ChannelAborted` in any situation where the channel is not accepted. This will fix the problem with "zombie channels" while still performing the rate limit check before the plugin.

To my mind it makes sense to do rate limiting upstream of running a plugin so that the plugin can depend on any channel being opened that it accepts.

See [024d78a](https://github.com/ACINQ/eclair/pull/2552/commits/024d78a8f7dc6b3cd8d57f7ed4cd8c486b2bcd67).

What do you think?

remyers replied:
Note: should I add a check to `PendingChannelRateLimiter` that rejects any attempt to open a channel with a `temporaryChannelId` that is already being tracked?

Otherwise someone could request a channel and then request another one with the same `temporaryChannelId` without it counting towards the rate limit. They could also fail the second request and remove both entries from `PendingChannelRateLimiter`. Generally, it's easier to reason about if we know there can be no duplicate channel ids.

t-bast replied:
Good idea, that makes sense!

t-bast replied:
> Note: should I add a check to PendingChannelRateLimiter that rejects any attempt to open a channel with a temporaryChannelId that is already being tracked?

No, that should be rejected even before that, and it should already be the case, the `Peer` is supposed to check that `temporaryChannelId` isn't duplicated.

remyers replied:
> > Note: should I add a check to PendingChannelRateLimiter that rejects any attempt to open a channel with a temporaryChannelId that is already being tracked?
> 
> No, that should be rejected even before that, and it should already be the case, the `Peer` is supposed to check that `temporaryChannelId` isn't duplicated.

Ha! Of course, that's pretty much all `Peer` does when it gets a new channel open request. I was over thinking things.

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala:
```diff
@@ -204,20 +181,25 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, wallet: OnchainP
+          }
+          stay() using d.copy(channels = d.channels + (TemporaryChannelId(temporaryChannelId) -> channel))
+        } else {
+          // we got disconnected or reconnected and this message was for the previous connection
```
I think the `ChannelClosed` event might not be the best choice because it requires parameters  that are not defined in `Peer`. If there were other actors watching for `ChannelClosed` they might do something unexpected if `Peer` emits a `ChannelEvent` produced with dummy values.

How about if `Peer` sends a message directly to `PendingChannelsRateLimiter` to remove a pending channel?

```scala
    log.warning(s"connection reset while waiting for open_channel to be accepted, ignoring")
    pendingChannelsRateLimiter ! RemoveChannelId(remoteNodeId, temporaryChannelId)
    stay()
```

Another option would be emit the new `ChannelAborted` message from PR #2593  .

t-bast replied:
You're right, `ChannelClosed` isn't the right choice here. But `ChannelAborted` should be exactly what we want!

remyers replied:
Fixed in [816284b](https://github.com/ACINQ/eclair/pull/2552/commits/816284b4a03ee9772a85583810217f50bc4aff4b). Will not emit `ChannelAborted` when a peer reconnects before `OpenChannelInterceptor` accepts the channel.

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala:
```diff
@@ -0,0 +1,108 @@
+  import PendingChannelsRateLimiter._
+
+  private def restoring(persistentChannelData: Seq[PersistentChannelData], pendingPeerChannels: Map[PublicKey, Seq[ByteVector32]], pendingPrivateNodeChannels: Seq[ByteVector32]): Behavior[Command] = {
+    persistentChannelData.headOption match {
```
Fixed in [dcd21b4](https://github.com/ACINQ/eclair/pull/2552/commits/dcd21b4962679312352689976014d672bd88f5e3). Good catch! I meant to do that but somehow slipped through the cracks.

I updated the tests to also include `normal` channels in the restored sequence to test the filter.

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala:
```diff
@@ -0,0 +1,108 @@
+
+  def apply(nodeParams: NodeParams, router: ActorRef[Any]): Behavior[Command] = {
+    Behaviors.setup { context =>
+      new PendingChannelsRateLimiter(nodeParams, router, context).restoring(nodeParams.db.channels.listLocalChannels(), Map(), Seq())
```
Fixed in [dcd21b4](https://github.com/ACINQ/eclair/pull/2552/commits/dcd21b4962679312352689976014d672bd88f5e3) to use and filter the loaded list of channels.

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala:
```diff
@@ -0,0 +1,293 @@
+
+  sealed trait WaitForRequestCommands extends Command
+  case class OpenChannelNonInitiator(remoteNodeId: PublicKey, open: Either[protocol.OpenChannel, protocol.OpenDualFundedChannel], temporaryChannelId: ByteVector32, fundingAmount: Satoshi, channelFlags: ChannelFlags, channelType_opt: Option[ChannelType], localFeatures: Features[InitFeature], remoteFeatures: Features[InitFeature], peerConnection: ActorRef[Any]) extends WaitForRequestCommands
+  case class OpenChannelInitiator(replyTo: ActorRef[Any], remoteNodeId: PublicKey, open: Peer.OpenChannel, localFeatures: Features[InitFeature], remoteFeatures: Features[InitFeature], peerConnection: ActorRef[Any]) extends WaitForRequestCommands
```
Fixed in [27e31c9](https://github.com/ACINQ/eclair/pull/2552/commits/27e31c9c3e38c20ab8dd1928c3e695b420f472f9)

remyers commented on eclair-core/src/main/resources/reference.conf:
```diff
@@ -140,6 +140,10 @@ eclair {
 
+  channel-opener-whitelist = [] // a list of public keys; we will ignore limits on pending channels from these peers
+  max-pending-channels-per-peer = 99 // maximum number of pending channels we will accept from a given peer
+  max-total-pending-channels-private-nodes = 99 // maximum number of pending channels we will accept from all private nodes
```
Fixed in [ab2b898](https://github.com/ACINQ/eclair/pull/2552/commits/ab2b8985d48bf08520c1a4cdde1c97700d7f2944) and [8b1afb4](https://github.com/ACINQ/eclair/pull/2552/commits/8b1afb42127c60119d7431c07373ffe7a785970b) (update release notes).

I agree that public nodes should not normally queue up a lot of pending channels, so 3 seems like a reasonable number.

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/PluginParams.scala:
```diff
@@ -55,4 +57,22 @@ trait CustomCommitmentsPlugin extends PluginParams {
+
+// @formatter:off
+trait InterceptOpenChannelCommand
+case class InterceptOpenChannelReceived(replyTo: ActorRef[InterceptOpenChannelResponse], openChannelNonInitiator: OpenChannelNonInitiator, temporaryChannelId: ByteVector32, defaultParams: DefaultParams) extends InterceptOpenChannelCommand {
```
fixed in [cc93ff5](https://github.com/ACINQ/eclair/pull/2552/commits/cc93ff5bc46287334fb5a645a238014d41ffe7c9#diff-986ea279ab7d792bcf6e6356d7e35ec5254648e9a28a8a77e65a883df4b9a58d)

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala:
```diff
@@ -404,33 +377,15 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, wallet: OnchainP
-      case None if Features.canUseFeature(localFeatures, remoteFeatures, Features.ChannelType) => Left(MissingChannelType(temporaryChannelId))
-      // remote doesn't specify a channel type: we use spec-defined defaults
-      case None => Right(ChannelTypes.defaultFromFeatures(localFeatures, remoteFeatures, channelFlags.announceChannel))
+  def handleOpenChannel(open: Either[protocol.OpenChannel, protocol.OpenDualFundedChannel], d: ConnectedData): Unit = {
```
Fixed in [cc93ff5](https://github.com/ACINQ/eclair/pull/2552/commits/cc93ff5bc46287334fb5a645a238014d41ffe7c9#diff-68459385a04ddc8b90528e792be9ee4f910fc7c9bd2a29c2903ec319832d13b1)

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala:
```diff
@@ -0,0 +1,108 @@
+        }
+      case None =>
+        context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelIdAssigned](c => ReplaceChannelId(c.remoteNodeId, c.temporaryChannelId, c.channelId)))
+        context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelOpened](c => RemoveChannelId(c.remoteNodeId, c.channelId)))
```
Don't forget to listen to both `ChannelAborted` and `ChannelClosed` as there are some transitions that could go directly to `CLOSED`

remyers replied:
Fixed in [43835bf](https://github.com/ACINQ/eclair/pull/2552/commits/43835bfe16bec1c0563c72c9dacce45d299cd719). Note also that I extended `ChannelClosed` from `ChannelEvents` in [fc8f1c8](https://github.com/ACINQ/eclair/pull/2552/commits/fc8f1c8b03c2ff21b83021f8f440aae0f9b10ba2).

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala:
```diff
@@ -0,0 +1,108 @@
+      case Some(d) =>
+        val adapter = context.messageAdapter[Router.GetNodeResponse](r => WrappedGetNodeResponse(d.channelId, r, None))
+        router ! GetNode(adapter, d.metaCommitments.remoteNodeId )
+        Behaviors.receiveMessagePartial[Command] {
```
I don't think the peer factory will be running before the pending channels rate limiter, so should not get any requests. To be safe though, it's fixed in [e10a214](https://github.com/ACINQ/eclair/pull/2552/commits/e10a2146419dd6413ee019167673b3949694b891).

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala:
```diff
@@ -0,0 +1,108 @@
+  case object ChannelRateLimited extends Response
+  // @formatter:on
+
+  def apply(nodeParams: NodeParams, router: ActorRef[Any]): Behavior[Command] = {
```
Oh, good! fixed in [aa90663](https://github.com/ACINQ/eclair/pull/2552/commits/aa9066391a52b0e194a65d2d8fea7992025cfc34).

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala:
```diff
@@ -0,0 +1,293 @@
+                           maxAcceptedHtlcs: Int)
+
+  def apply(peer: ActorRef[Any], nodeParams: NodeParams, wallet: OnchainPubkeyCache, pendingChannelsRateLimiter: ActorRef[PendingChannelsRateLimiter.Command], pluginTimeout: FiniteDuration = 1 minute): Behavior[Command] =
+    Behaviors.setup {
```
Fixed in [28e7bc2](https://github.com/ACINQ/eclair/pull/2552/commits/28e7bc25d6d595523e616dd6b5ff377defb33763).

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala:
```diff
@@ -0,0 +1,108 @@
+    }
+  }
+
+  private def registering(pendingPeerChannels: Map[PublicKey, Seq[ByteVector32]], pendingPrivateNodeChannels: Seq[ByteVector32]): Behavior[Command] = {
```
Added monitoring in [af22754](https://github.com/ACINQ/eclair/pull/2552/commits/af2275477cab1da6eca8a44d0d399d3a1f8e1076).

Should we also track how often channel pending requests end up as opened/closed/aborted ?

t-bast replied:
> Should we also track how often channel pending requests end up as opened/closed/aborted ?

I don't think it's very useful, so let's ignore it for now. We can add it later if we feel it's necessary.

remyers commented:
An example channel interceptor plugin is proposed in [eclair-plugins PR #4](https://github.com/ACINQ/eclair-plugins/pull/4) .

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala:
```diff
@@ -0,0 +1,276 @@
+    }
+  }
+
+  private def sanityCheckInitiator(initiator: OpenChannelInitiator): Behavior[Command] = {
```
nit: it's a bit confusing that you call this variable `initiator` (and in the other case `nonInitiator`), that's not the important meaning behind the variable, it's rather the `OpenChannel` part that is the most important...it makes more sense to call these variables `request` or `cmd` in both cases in my opinion.

remyers replied:
Fixed in [a032b09](https://github.com/ACINQ/eclair/pull/2552/commits/a032b097a5c1ccb8e04985f7714cc6fbd2937663).

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala:
```diff
@@ -0,0 +1,276 @@
+    pendingChannelsRateLimiter ! AddOrRejectChannel(adapter, nonInitiator.remoteNodeId, nonInitiator.temporaryChannelId)
+    receiveCommandMessage[CheckRateLimitsCommands](context, "checkRateLimits") {
+      case PendingChannelsRateLimiterResponse(PendingChannelsRateLimiter.AcceptOpenChannel) =>
+        validateRemoteChannelType(nonInitiator.temporaryChannelId, nonInitiator.channelFlags, nonInitiator.channelType_opt, nonInitiator.localFeatures, nonInitiator.remoteFeatures) match {
```
I think it would make more sense to start with a `sanityCheckNonInitiator()` first to run all _synchronous_ checks, and afterwards run the _asynchronous_ checks (rate-limiting and plugin).

remyers replied:
Good idea; fixed in [3c4137f](https://github.com/ACINQ/eclair/pull/2552/commits/3c4137fdb0bb6c8f712c40f508c9fc0468c13304).

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala:
```diff
@@ -0,0 +1,276 @@
+      val defaultParams = DefaultParams(localParams.dustLimit, localParams.maxHtlcValueInFlightMsat, localParams.htlcMinimum, localParams.toSelfDelay, localParams.maxAcceptedHtlcs)
+      plugin.openChannelInterceptor ! InterceptOpenChannelReceived(pluginResponseAdapter, nonInitiator, defaultParams)
+      receiveCommandMessage[QueryPluginCommands](context, "queryPlugin") {
+        case PluginOpenChannelResponse(pluginResponse: AcceptOpenChannel) =>
+          val localParams1 = updateLocalParams(localParams, pluginResponse.defaultParams)
+          peer ! SpawnChannelNonInitiator(nonInitiator.open, channelConfig, channelType, localParams1, nonInitiator.peerConnection.toClassic)
+          waitForRequest()
+        case PluginOpenChannelResponse(pluginResponse: RejectOpenChannel) =>
+          sendFailure(PluginRejectedError(pluginResponse.error), nonInitiator)
+          waitForRequest()
```
You should cancel the timer in both cases, otherwise this will spam the logs with "received unhandled message: PluginTimeout" errors.

remyers replied:
Fixed in [3c4137f](https://github.com/ACINQ/eclair/pull/2552/commits/3c4137fdb0bb6c8f712c40f508c9fc0468c13304).

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/OpenChannelInterceptor.scala:
```diff
@@ -0,0 +1,276 @@
+  private case class PluginOpenChannelResponse(pluginResponse: InterceptOpenChannelResponse) extends QueryPluginCommands
+  private case object PluginTimeout extends QueryPluginCommands
+
+  private sealed trait ErrorResponse
+  private object RateLimitError extends ErrorResponse {
+    override def toString: String = "rate limit reached"
+  }
+  private case class ChannelExceptionError(ex: ChannelException) extends ErrorResponse {
+    override def toString: String = ex.getMessage
+  }
+
+  private object ConcurrentRequestError extends ErrorResponse {
+    override def toString: String = "concurrent request rejected"
+  }
+  private object PluginTimeoutError extends ErrorResponse {
+    override def toString: String = "plugin timeout"
+  }
+  private case class PluginRejectedError(error: Error) extends ErrorResponse {
+    override def toString: String = error.toAscii
+  }
```
Each of these errors is used exactly once, and only used for its `toString` value. This is an indication that it's probably overly verbose to define this trait, I would remove them completely and just change the signature of `sendFailure` to:

```
private def sendFailure(error: String, nonInitiator: OpenChannelNonInitiator): Unit
```

In each error case, you would just inline the error message.

remyers replied:
fixed in [3c4137f](https://github.com/ACINQ/eclair/pull/2552/commits/3c4137fdb0bb6c8f712c40f508c9fc0468c13304).

t-bast commented on eclair-core/src/test/scala/fr/acinq/eclair/io/PeerSpec.scala:
```diff
@@ -91,9 +91,16 @@ class PeerSpec extends FixtureSpec {
     }
 
-    val peer: TestFSMRef[Peer.State, Peer.Data, Peer] = TestFSMRef(new Peer(aliceParams, remoteNodeId, wallet, FakeChannelFactory(channel), switchboard.ref))
+    object MockLimiter extends Actor {
+      def receive: Receive = {
+        case msg: PendingChannelsRateLimiter.AddOrRejectChannel => msg.replyTo ! PendingChannelsRateLimiter.AcceptOpenChannel
+      }
+    }
+    val mockLimiter = system.actorOf(Props(MockLimiter)).ref
```
You should use a `TestProbe` on which you `setAutopilot` for this kind of use-case. You should be able to find examples of this across the codebase.

remyers replied:
Fixed in [5a7b110](https://github.com/ACINQ/eclair/pull/2552/commits/5a7b11076a9de8f644b143755384df1abe4b000c).

t-bast commented on eclair-core/src/test/scala/fr/acinq/eclair/wire/internal/channel/ChannelCodecsSpec.scala:
```diff
@@ -334,4 +334,34 @@ object ChannelCodecsSpec {
     DATA_NORMAL(MetaCommitments(commitments), ShortIds(RealScidStatus.Final(RealShortChannelId(42)), ShortChannelId.generateLocalAlias(), None), None, channelUpdate, None, None, None)
   }
 
+  def makeChannelDataPending(remoteNodeId: Crypto.PublicKey, channelId: ByteVector32): DATA_WAIT_FOR_CHANNEL_READY = {
+    val fundingTx = Transaction.read("0200000001adbb20ea41a8423ea937e76e8151636bf6093b70eaff942930d20576600521fd000000006b48304502210090587b6201e166ad6af0227d3036a9454223d49a1f11839c1a362184340ef0240220577f7cd5cca78719405cbf1de7414ac027f0239ef6e214c90fcaab0454d84b3b012103535b32d5eb0a6ed0982a0479bbadc9868d9836f6ba94dd5a63be16d875069184ffffffff028096980000000000220020c015c4a6be010e21657068fc2e6a9d0 ...
+    val fundingAmount = fundingTx.txOut.head.amount
+    val commitmentInput = Funding.makeFundingInputInfo(fundingTx.hash, 0, fundingAmount, channelKeyManager.fundingPublicKey(localParams.fundingKeyPath).publicKey, remoteParams.fundingPubKey)
+
+    val remoteSig = ByteVector64(hex"2148d2d4aac8c793eb82d31bcf22d4db707b9fd7eee1b89b4b1444c9e19ab7172bab8c3d997d29163fa0cb255c75afb8ade13617ad1350c1515e9be4a222a04d")
+    val commitTx = Transaction(
+      version = 2,
+      txIn = TxIn(
+        outPoint = commitmentInput.outPoint,
+        signatureScript = ByteVector.empty,
+        sequence = 0,
+        witness = Scripts.witness2of2(randomBytes64(), remoteSig, channelKeyManager.fundingPublicKey(localParams.fundingKeyPath).publicKey, remoteParams.fundingPubKey)) :: Nil,
+      txOut = Nil,
+      lockTime = 0
+    )
+    val localCommit = LocalCommit(0, CommitmentSpec(Set(), FeeratePerKw(1500 sat), 50000000 msat, 70000000 msat), CommitTxAndRemoteSig(CommitTx(commitmentInput, commitTx), remoteSig), Nil)
+    val remoteCommit = RemoteCommit(0, CommitmentSpec(Set(), FeeratePerKw(1500 sat), 50000 msat, 700000 msat), ByteVector32(hex"0303030303030303030303030303030303030303030303030303030303030303"), PrivateKey(ByteVector.fill(32)(4)).publicKey)
+    val commitments = Commitments(channelId, ChannelConfig.standard, ChannelFeatures(), localParams, remoteParams.copy(nodeId = remoteNodeId), ChannelFlags.Public, localCommit, remoteCommit, LocalChanges(Nil, Nil, Nil), RemoteChanges(Nil, Nil, Nil),
+      localNextHtlcId = 32L,
+      remoteNextHtlcId = 4L,
+      originChannels = Map(),
+      remoteNextCommitInfo = Right(randomKey().publicKey),
+      localFundingStatus = UnknownFundingTx,
+      remoteFundingStatus = RemoteFundingStatus.NotLocked,
+      remotePerCommitmentSecrets = ShaChain.init)
+
+    DATA_WAIT_FOR_CHANNEL_READY(MetaCommitments(commitments), ShortIds(RealScidStatus.Final(RealShortChannelId(42)), ShortChannelId.generateLocalAlias(), None))
+  }
```
I don't think this is necessary, you should be able to just use `CommitmentsSpec.makeCommitments` to get a `MetaCommitments` object, and from this it should be trivial to create some `DATA_WAIT_FOR_FUNDING_CONFIRMED`, `DATA_WAIT_FOR_CHANNEL_READY`, `DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED` and `DATA_WAIT_FOR_DUAL_FUNDING_READY` for test cases.

You only care about setting the `remoteNodeId` and reading the `channelId` in your tests, right?

remyers replied:
I rolled back my changes to `ChannelCodecsSpec.scala` and replaced it with a simpler `makeChannelDataPending`  function local to the tests that use it and create the metaCommitments as you suggested. See [b4c3ba0](https://github.com/ACINQ/eclair/pull/2552/commits/b4c3ba0146aa599d3642d417e18c4d1547c41179).

Do you think I should create tests for each of the four pending states that are filtered for when you restore pending requests from the channel db? Right now I use `DATA_WAIT_FOR_DUAL_FUNDING_READY` as a proxy for the others since the code is so simple there.

t-bast reviewed (commented):
Mostly nit comments, I've now reviewed everything except the `PendingChannelsRateLimiterSpec` and `OpenChannelInterceptorSpec`, I'll have a look at them shortly.

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala:
```diff
@@ -0,0 +1,126 @@
+        }
+      case RemoveChannelId(remoteNodeId, channelId) =>
+        pendingPeerChannels.get(remoteNodeId) match {
+          case Some(pendingChannels) => registering((pendingPeerChannels + (remoteNodeId -> pendingChannels.filterNot(_ == channelId))).filter(_._2.nonEmpty), pendingPrivateNodeChannels)
```
perhaps also more efficient to use if vs filter; fixed in [70b19a6](https://github.com/ACINQ/eclair/pull/2552/commits/70b19a6bfff7751050401ce5fd7b4ab80dfdcc17)

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala:
```diff
@@ -0,0 +1,126 @@
+
+  private def restoring(channels: Map[PublicKey, Seq[PersistentChannelData]], pendingPeerChannels: Map[PublicKey, Seq[ByteVector32]], pendingPrivateNodeChannels: Seq[ByteVector32]): Behavior[Command] = {
+    channels.headOption match {
+      case Some(d) =>
```
fixed in [d7ce81c](https://github.com/ACINQ/eclair/pull/2552/commits/d7ce81c7b6402062ac8be8e63d1bd2f451851bf7)

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala:
```diff
@@ -0,0 +1,126 @@
+  private def registering(pendingPeerChannels: Map[PublicKey, Seq[ByteVector32]], pendingPrivateNodeChannels: Seq[ByteVector32]): Behavior[Command] = {
+    Metrics.OpenChannelRequestsPending.withTag(Tags.PublicPeers, value = true).update(pendingPeerChannels.flatMap(_._2).size)
+    Metrics.OpenChannelRequestsPending.withTag(Tags.PublicPeers, value = false).update(pendingPrivateNodeChannels.size)
+    Behaviors.receiveMessagePartial[Command] {
```
fixed in [d7ce81c](https://github.com/ACINQ/eclair/pull/2552/commits/d7ce81c7b6402062ac8be8e63d1bd2f451851bf7)

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala:
```diff
@@ -0,0 +1,126 @@
+            registering(pendingPeerChannels + (announcement.nodeId -> (temporaryChannelId +: peerChannels)), pendingPrivateNodeChannels)
+          case None =>
+            replyTo ! AcceptOpenChannel
+            registering(Map(announcement.nodeId -> Seq(temporaryChannelId)), pendingPrivateNodeChannels)
```
ouch! fixed and tested in [da3d570](https://github.com/ACINQ/eclair/pull/2552/commits/da3d5701cb10b9be25624ea7eb0b081fd1cf9428).

remyers commented on docs/release-notes/eclair-vnext.md:
```diff
@@ -27,6 +27,28 @@ Eclair offers three strategies to handle that scenario, that node operators can
+channel.channel-open-limits.max-pending-channels-per-peer = 3 
+
+// maximum number of pending channels we will accept from all private nodes
+channel.channel-open-limits.max-total-pending-channels-private-nodes = 99 
```
fixed in [2e8bbd3](https://github.com/ACINQ/eclair/pull/2552/commits/2e8bbd30fc8eec68b665d3d9d0c1309d4cce680c)

t-bast commented on eclair-core/src/test/scala/fr/acinq/eclair/io/OpenChannelInterceptorSpec.scala:
```diff
@@ -0,0 +1,153 @@
+import scala.concurrent.duration.DurationInt
+
+class OpenChannelInterceptorSpec extends ScalaTestWithActorTestKit(ConfigFactory.load("application")) with FixtureAnyFunSuiteLike {
+  val fundingAmount: Satoshi = 1 sat
+  val publicKey: Crypto.PublicKey = PrivateKey(ByteVector32.One).publicKey
+  val remoteNodeId: Crypto.PublicKey = PrivateKey(ByteVector32.One).publicKey
+  val temporaryChannelId: ByteVector32 = ByteVector32.Zeroes
+  val openChannel: OpenChannel = OpenChannel(ByteVector32.Zeroes, temporaryChannelId, fundingAmount, 0 msat, 1 sat, UInt64(1), 1 sat, 1 msat, FeeratePerKw(1 sat), CltvExpiryDelta(1), 1, publicKey, publicKey, publicKey, publicKey, publicKey, publicKey, ChannelFlags.Private)
+  val localParams: LocalParams = LocalParams(remoteNodeId, DeterministicWallet.KeyPath(Seq(42L)), 1 sat, Long.MaxValue.msat, Some(500 sat), 1 msat, CltvExpiryDelta(144), 50, isInitiator = false, None, None, Features.empty)
+  val defaultParams: DefaultParams = DefaultParams(100 sat, 100000 msat, 100 msat, CltvExpiryDelta(288), 10)
+  val channels: Map[Peer.FinalChannelId, actor.ActorRef] = Map(Peer.FinalChannelId(randomBytes32()) -> system.classicSystem.deadLetters)
+  val connectedData: Peer.ConnectedData = Peer.ConnectedData(NodeAddress.fromParts("1.2.3.4", 42000).get, system.classicSystem.deadLetters, Init(TestConstants.Alice.nodeParams.features.initFeatures()), Init(TestConstants.Bob.nodeParams.features.initFeatures()), channels.map { case (k: ChannelId, v) => (k, v) })
+  def publicKey(fill: Byte): Crypto.PublicKey = PrivateKey(ByteVector.fill(32)(fill)).publicKey
```
Most of this code is unused, I think you only need the following:

```suggestion
val remoteNodeId: Crypto.PublicKey = randomKey().publicKey
  val openChannel: OpenChannel = OpenChannel(ByteVector32.Zeroes, randomBytes32(), 250_000 sat, 0 msat, 1 sat, UInt64(1), 1 sat, 1 msat, FeeratePerKw(1 sat), CltvExpiryDelta(1), 1, randomKey().publicKey, randomKey().publicKey, randomKey().publicKey, randomKey().publicKey, randomKey().publicKey, randomKey().publicKey, ChannelFlags.Private)
  val defaultParams = DefaultParams(100 sat, 100_000 msat, 100 msat, CltvExpiryDelta(288), 10)
```

In particular, the `connectedData` was only used to access `localFeatures` and `remoteFeatures`. We should instead let each test specify what features are used: in most cases it will just be `Features.empty` for both as it has no impact on the test, except when you want to test that dual funding is used, where you'll explicitly add the dual funding feature in the corresponding tests.

remyers replied:
Good point! in fact, `openChannel` can be replaced by `createOpenChannelMessage`. Fixed in [522019b](https://github.com/ACINQ/eclair/pull/2552/commits/522019bb22f2ff980a173707db0ad7731583054b).

remyers commented on eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala:
```diff
@@ -0,0 +1,364 @@
+    DATA_WAIT_FOR_FUNDING_CONFIRMED(commitments, BlockHeight(0), None, Left(FundingCreated(channelId, ByteVector32.Zeroes, 3, randomBytes64())))
+  }
+
+  case class FixtureParam(router: TestProbe[Router.GetNode], nodeParams: NodeParams, probe: TestProbe[PendingChannelsRateLimiter.Response], limiter: ActorRef[PendingChannelsRateLimiter.Command], eventListener: TestProbe[ChannelEvent])
```
removed in [48889ac](https://github.com/ACINQ/eclair/pull/2552/commits/48889acfb76add200a7e61e4d6bd0bc5d328603d)

remyers commented on eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala:
```diff
@@ -0,0 +1,364 @@
+    system.eventStream ! Subscribe(eventListener.ref)
+
+    withFixture(test.toNoArgTest(FixtureParam(router, nodeParams, probe, limiter, eventListener)))
+  }
```
I've reworked the tests to follow more of an actual usage flow and with your suggested fixture changes in [48889ac](https://github.com/ACINQ/eclair/pull/2552/commits/48889acfb76add200a7e61e4d6bd0bc5d328603d).

In the process I found an additional bug, which I fixed in [6fa126c](https://github.com/ACINQ/eclair/pull/2552/commits/6fa126c376e0a2838275f215537bcf52c36dc0b5).

If I roll back either [6fa126c](https://github.com/ACINQ/eclair/pull/2552/commits/6fa126c376e0a2838275f215537bcf52c36dc0b5) or [da3d570](https://github.com/ACINQ/eclair/pull/2552/commits/da3d5701cb10b9be25624ea7eb0b081fd1cf9428), these tests fail where expected.

Do you think these tests look easier to understand/review?

remyers commented on eclair-core/src/test/scala/fr/acinq/eclair/io/PeerSpec.scala:
```diff
@@ -52,7 +53,7 @@ class PeerSpec extends FixtureSpec {
   override implicit val patienceConfig: PatienceConfig = PatienceConfig(timeout = 30 seconds, interval = 1 second)
 
-  case class FixtureParam(nodeParams: NodeParams, remoteNodeId: PublicKey, system: ActorSystem, peer: TestFSMRef[Peer.State, Peer.Data, Peer], peerConnection: TestProbe, channel: TestProbe, switchboard: TestProbe) {
+  case class FixtureParam(nodeParams: NodeParams, remoteNodeId: PublicKey, system: ActorSystem, peer: TestFSMRef[Peer.State, Peer.Data, Peer], peerConnection: TestProbe, channel: TestProbe, switchboard: TestProbe, mockLimiter: ActorRef) {
```
Missing tests added in [7958e24](https://github.com/ACINQ/eclair/pull/2552/commits/7958e249faf0a45eb7ad1bb438b9647ce60524d4). 

Rejecting duplicate `temporaryChannelId`s is already implemented.

t-bast reviewed (commented):
Even though they're not trivial to read, I believe the tests are much better now (especially since they allowed finding another bug). I added some cleanup in https://github.com/ACINQ/eclair/commit/8ed8169da58c873c38772dc387c40858206a95cd that you may want to apply.

Looking at these updates tests made me find another issue. We need to change `pendingPrivateNodeChannels` to be a `Map[PublicKey, Seq[ByteVector32]]` like `pendingPeerChannels`, otherwise an attacker can do the following:

- send open_channel with **different** node_ids but with the **same** temporary_channel_id until they hit the limit
- we then have maxTotalPendingChannelsPrivateNodes channels pending
- they abort one of those channels, which emits a ChannelAborted event
- but that makes us empty the pendingPrivateNodeChannels since it contains the same temporary_channel_id duplicated maxTotalPendingChannelsPrivateNodes times
- which means they can re-do step 1 and add another batch of maxTotalPendingChannelsPrivateNodes pending channels
- they effectively bypassed our rate-limits

The actual identifier for a channel is the pair <node_id, channel_id>, even for channels with private nodes. We should start by updating the "requests from private nodes are only accepted and tracked while under global limit" unit test with a duplicate temporary_channel_id reused by two distinct node_ids, verify that it fails, then fix the `PendingChannelsRateLimiter` implementation.

remyers commented:
> Even though they're not trivial to read, I believe the tests are much better now (especially since they allowed finding another bug). I added some cleanup in [8ed8169](https://github.com/ACINQ/eclair/commit/8ed8169da58c873c38772dc387c40858206a95cd) that you may want to apply.

That looks good, I added one fixup in [4760467](https://github.com/ACINQ/eclair/pull/2552/commits/4760467825f8e6973e68a08bfdbd6df671925529) so that all three events (confirmed/closed/aborted) are covered by using both `peersBelowLimit` nodes tested in the next code block.

remyers commented:
> The actual identifier for a channel is the pair <node_id, channel_id>, even for channels with private nodes. We should start by updating the "requests from private nodes are only accepted and tracked while under global limit" unit test with a duplicate temporary_channel_id reused by two distinct node_ids, verify that it fails, then fix the `PendingChannelsRateLimiter` implementation.

The test is fixed (and now fails) in [f60f154](https://github.com/ACINQ/eclair/pull/2552/commits/f60f154a53dfdd16c097642d1e81a5939cdba9cd) and `PendingChannelsRateLimiter` is fixed (and the test succeeds) in the next commit, [a26d41c](https://github.com/ACINQ/eclair/pull/2552/commits/a26d41c8638294b82e1d050f13a910a0e642f13a).

remyers commented on eclair-core/src/main/scala/fr/acinq/eclair/PluginParams.scala:
```diff
@@ -55,4 +57,20 @@ trait CustomCommitmentsPlugin extends PluginParams {
+// @formatter:on
+
+trait InterceptOpenChannelPlugin extends PluginParams {
+  def openChannelInterceptor: ActorRef[InterceptOpenChannelCommand]
```
I just noticed I used a `def` here, but it should be a `val`. Right?

t-bast replied:
No, you should never use `val` in a `trait`, it should always be a `def` since the `trait` doesn't know how the concrete classes will implement this. The concrete class that implements this `trait` will most likely use a `val` though.

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala:
```diff
@@ -0,0 +1,154 @@
+            registering(pendingPublicNodeChannels + (announcement.nodeId -> Seq(temporaryChannelId)), pendingPrivateNodeChannels)
+        }
+      case WrappedGetNodeResponse(temporaryChannelId, UnknownNode(nodeId), Some(replyTo)) =>
+        if (pendingPrivateNodeChannels.flatMap(_._2).size >= nodeParams.channelConf.maxTotalPendingChannelsPrivateNodes) {
```
nit: this creates a new list from each private node's channels list just to get the size, which is quite wasteful. The following code should be better:

```suggestion
if (pendingPrivateNodeChannels.map(_._2.length).sum >= nodeParams.channelConf.maxTotalPendingChannelsPrivateNodes) {
```

remyers replied:
Good idea; it's fixed in [9b27eed](https://github.com/ACINQ/eclair/pull/2552/commits/9b27eed4003dde3e8440f43a3d87bd548f95b058)

t-bast reviewed (approved):
LGTM :100:

pm47 commented:
It looks that one test is flaky: https://github.com/ACINQ/eclair/actions/runs/4233344581/jobs/7354182704.

t-bast commented:
@remyers can you investigate that?

remyers commented:
> @remyers can you investigate that?

I suspect the three published events prior to this failure have (sometimes) not all been processed by the `limiter` actor before it is queried. Is there a recommended way to detect when all events have been received by an actor under test? 

I don't think this will work because I still haven't ensured the `limiter` has seen the published `ChannelEvent`s:
```scala
    val eventListener = TestProbe[ChannelEvent]()
    system.eventStream ! Subscribe(eventListener.ref)
    system.eventStream ! Publish(ChannelOpened(null, peersAtLimit.head, channelIdAtLimit1))
    system.eventStream ! Publish(ChannelClosed(null, newChannelId1, null, commitments(peersBelowLimit.head, newChannelId1)))
    system.eventStream ! Publish(ChannelAborted(null, peersBelowLimit.last, newChannelId2))
    eventListener.expectMessageType[ChannelOpened]
    eventListener.expectMessageType[ChannelClosed]
    eventListener.expectMessageType[ChannelAborted] 

    // how can we know when `limiter` has processed the last event before calling
    limiter ! PendingChannelsRateLimiter.AddOrRejectChannel(probe.ref, peer, randomBytes32()) 
```
